crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -14,6 +14,10 @@ use crate::prelude::*;
 
 #[derive(Eq, PartialEq, Debug, Default)]
 pub enum TupleParentheses {
+    /// TODO(cnpryer): Why not `Never`?
+    /// Black omits parentheses for tuples inside of comprehensions.
+    Comprehension,
```

Comment:
How does `Comprehension` fit into your other PR where you tried to generalize the parentheses rule names (would this match `Never`)? Is it just comprehension where tuples aren't wrapped even if they exceed the line width?

Reply:
This PR predates that PR. I wasn't sure if `Comprehension` matched the pattern that was already materializing or there was room to improve. I decided to propose an alternative that I'll try to shortly explain below.

> Is it just comprehension where tuples aren't wrapped even if they exceed the line width?

I kinda wanna answer this with another question. From what perspective is `TupleParentheses` useful?

To me it's useful in defining formatting behavior for a tuple's parentheses. So if I'm implementing `Comprehension` I'd want to say "tuples here don't have parentheses" I'd probably use something closer to `TupleParentheses::Never`.

But to add to this, I'm still curious of understanding `NeedsParentheses`' role in all of this.

Is there a way to incorporate some of this configuration in how `NeedsParentheses` is implemented?

```rs
impl NeedsParentheses for ExprTuple {
    fn needs_parentheses(
        &self,
        parent: AnyNodeRef,
        _context: &PyFormatContext,
    ) -> OptionalParentheses {
        if parent.is_comprehension() {
            ...
        }
    }
}
```

Does this imply that if `needs_parentheses` is invoked somewhere we could just handle it with knowledge of the tuple's `parent`?

Is this where `TupleParentheses::Expr(Parentheses::Never)` becomes useful?

Reply:
What do you think about merging with `TupleParentheses::Comprehension` here and then I can better document motivations in #5810.

Reply:
> But to add to this, I'm still curious of understanding NeedsParentheses' role in all of this.

Okay, let's see if I can explain this well. 

It is possible to parenthesize an expression in Python (and most other languages). Adding parentheses allows to express the precedence explicitly. All these are parenthesized expressions: `(a)`, `(1), `([a, b])`.

Tuples are special because they have their own parentheses (like lists or dicts too), but they also happen to be the same as for parenthesizing expressions. The tuple parentheses are, in some situations, optional: so you can write `a, b` or `(a, b)`. The tuple `(a, b)` is a tuple with its optional parentheses, but it is not a parenthesized expression. It is only a parenthesized expression if you add one extra pair of parentheses: `((a, b))`. 

The logic in `FormatExpr` and `NeedsParantheses` is only concerned about the "generic" parentheses (parenthesized expressions). That's why `TupleParentheses` is the right approach in my view, because it doesn't control whether the expression should be parenthesized, but whether the tuple should preserve its parentheses. 

> Is this where TupleParentheses::Expr(Parentheses::Never) becomes useful?

I don't think this variant is used anymore and we should not use it to determine whether to keep the tuple parentheses (except if we **always** must preserve them if the expression is parenthesized). I must have overlooked it by one of my refactors that removed the `Parenthesize::Custom` variant.

> What do you think about merging with TupleParentheses::Comprehension here and then I can better document motivations in #5810.

Sounds good to me. Can you remove the `TODO Never` before merging?

Reply:
This is so awesome. Thank you for all the detail. Happy to curate some docs on it if I can too.

3ddff0b9f51beaf88cd1b3907cc4c82f98c5d908

Reply:
> removed the `Parentheses::Custom`

#5708 right? I'm checking that out now.

Reply:
Yes, that's the PR that I had in mind but it didn't change `TupleParentheses::Expr` (or any logic related to it). It seems, that this variant might have been unused from the beginning